File: LeetCode/python/lc_0076.py
"""
 * @file     lc_0076.py
 * @brief    https://leetcode-cn.com/problems/minimum-window-substring/
 * @author   YongDu
 * @date     2021-11-01
"""


class Solution:
    def minWindow(self, s: str, t: str) -> str:
        dst_map = defaultdict(int)
        dst_len = len(t)

        for ch in t:
            dst_map[ch] += 1

        min_left, min_right = 0, len(s)
        left = 0

        for right, ch in enumerate(s):
            if dst_map[ch] > 0:
                dst_len -= 1
            dst_map[ch] -= 1

            if dst_len == 0:
                # 收缩窗口
                while dst_map[s[left]] < 0:
                    dst_map[s[left]] += 1
                    left += 1

                # 记录最小窗口
                if right - left < min_right - min_left:
                    min_left, min_right = left, right

                # 扩大窗口
                dst_map[s[left]] += 1
                dst_len += 1
                left += 1

        return "" if min_right == len(s) else s[min_left:min_right + 1]

File: LeetCode/cpp/lc_0076.cpp
/**
 * @file     lc_0076.cpp
 * @brief    https://leetcode-cn.com/problems/minimum-window-substring/
 * @author   YongDu
 * @date     2021-09-07
 */
class Solution {
    using charMap = std::unordered_map<char, int>;

  public:
    string minWindow(string s, string t) {
        charMap dstMap;
        charMap curMap;

        for (auto& ch : t) {
            dstMap[ch]++;
        }

        int minLen = INT_MAX; // 最小长度
        int leftBound = -1;   // 最小长度左边界，最后截取子串
        int left = 0;
        for (int right = 0; right < s.size(); ++right) {
            curMap[s[right]]++;
            while (isValidSubStr(curMap, dstMap) && left <= right) {
                if (right - left + 1 < minLen) {
                    minLen = right - left + 1;
                    leftBound = left;
                }
                --curMap[s[left]];
                left++;
            }
        }
        return minLen == INT_MAX ? string{} : s.substr(leftBound, minLen);
    }

  private:
    bool isValidSubStr(charMap& curMap, charMap& dstMap) {
        for (auto& [ch, num] : dstMap) {
            if (curMap[ch] < num) {
                return false;
            }
        }
        return true;
    }
};
